internal/mage/engine.go

```diff
@@ -48,3 +52,165 @@ func (t Engine) Lint(ctx context.Context) error {
 		}).ExitCode(ctx)
 	return err
 }
+
+const (
+	sdkHelper      = "dagger-sdk-helper"
+	buildkitRepo   = "github.com/moby/buildkit"
+	buildkitBranch = "v0.10.5"
+	// TODO: placeholder until real one exists
+	engineImageRef = "eriksipsma/dagger-test:bootstrap"
+)
+
+func engineContainer(c *dagger.Client, arches, oses []string) []*dagger.Container {
+	buildkitRepo := c.Git(buildkitRepo).Branch(buildkitBranch).Tree()
+
+	var platformVariants []*dagger.Container
+	for _, arch := range arches {
+		buildkitBase := c.Container(dagger.ContainerOpts{
+			Platform: dagger.Platform("linux/" + arch),
+		}).Build(buildkitRepo)
+		for _, os := range oses {
+			// include each helper for each arch too in case there is a
+			// client/server mismatch
+			for _, arch := range arches {
+				helperBin := util.GoBase(c).
+					WithEnvVariable("GOOS", os).
+					WithEnvVariable("GOARCH", arch).
+					Exec(dagger.ContainerExecOpts{
+						Args: []string{"go", "build", "-o", "./bin/" + sdkHelper, "-ldflags", "-s -w", "/app/cmd/sdk-helper"},
+					}).
+					File("./bin/" + sdkHelper)
+				buildkitBase = buildkitBase.WithFS(
+					buildkitBase.FS().WithFile("/usr/bin/"+sdkHelper+"-"+os+"-"+arch, helperBin),
+				)
+			}
+		}
+		platformVariants = append(platformVariants, buildkitBase)
+	}
+
+	return platformVariants
+}
+
+func (t Engine) Release(ctx context.Context) error {
+	c, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stderr))
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	arches := []string{"amd64", "arm64"}
+	oses := []string{"linux", "darwin"}
+
+	imageRef, err := c.Container().Publish(ctx, engineImageRef, dagger.ContainerPublishOpts{
+		PlatformVariants: engineContainer(c, arches, oses),
+	})
+	if err != nil {
+		return err
+	}
+	fmt.Println("Image published:", imageRef)
+
+	return nil
+}
+
+func localEngine(ctx context.Context, c *dagger.Client) (string, error) {
+	tmpfile, err := os.CreateTemp("", "dagger-engine-export")
+	if err != nil {
+		return "", err
+	}
+	defer os.Remove(tmpfile.Name())
+
+	_, err = c.Container().Export(ctx, tmpfile.Name(), dagger.ContainerExportOpts{
+		PlatformVariants: engineContainer(c, []string{runtime.GOARCH}, []string{runtime.GOOS}),
```

Comment:
I think this won't work on Mac

Reply:
It works, just double checked again on my mac directly. `engineContainer` always uses `linux` as the base OS (since that's all we support), specifying `darwin` as an os here only results in darwin helper binaries being include in the image.